File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/fwmengine/world/ecs/systems/PathSystem.java
package de.limbusdev.guardianmonsters.fwmengine.world.ecs.systems;

import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.EntitySystem;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.utils.ImmutableArray;
import com.badlogic.gdx.utils.TimeUtils;

import de.limbusdev.guardianmonsters.Constant;
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.ColliderComponent;
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.Components;
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.InputComponent;
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.PathComponent;
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.PositionComponent;
import de.limbusdev.utils.geometry.IntRect;
import de.limbusdev.utils.geometry.IntVec2;


/**
 * Moves around entities with a {@link PathComponent} like persons, animals and so on
 * Created by georg on 30.11.15.
 */
public class PathSystem extends EntitySystem {
    /* ............................................................................ ATTRIBUTES .. */
    private ImmutableArray<Entity> entities;
    private GameArea gameArea;
    /* ........................................................................... CONSTRUCTOR .. */
    public PathSystem(GameArea gameArea) {
        this.gameArea = gameArea;
    }
    /* ............................................................................... METHODS .. */
    public void addedToEngine(Engine engine) {
        entities = engine.getEntitiesFor(Family.all(
                PositionComponent.class,
                ColliderComponent.class,
                PathComponent.class).exclude(
                InputComponent.class
        ).get());
    }

    public void update(float deltaTime) {
        for (Entity entity : entities) {
            PositionComponent position = Components.getPositionComponent(entity);
            ColliderComponent collider = Components.getColliderComponent(entity);
            PathComponent path = Components.getPathComponent(entity);
            makeOneStep(position, path, collider);
            position.updateGridPosition();
        }
    }

    public void makeOneStep(PositionComponent position, PathComponent path, ColliderComponent
            collider) {
        if(path.startMoving && !path.staticEntity) {
            // Define direction of movement
            switch(path.path.get(path.currentDir)) {
                case N: position.nextX= position.getX();position.nextY = position.getY() + Constant.TILE_SIZE;
                    break;
                case W: position.nextX= position.getX() - Constant.TILE_SIZE;position.nextY = position.getY();
                    break;
                case E: position.nextX= position.getX() + Constant.TILE_SIZE;position.nextY = position.getY();
                    break;
                case S: position.nextX= position.getX();position.nextY = position.getY() - Constant.TILE_SIZE;
                    break;
                default: position.nextY= position.getX();position.nextY= position.getY();break;
            }

            /* Check whether movement is possible or blocked by a collider */
            IntVec2 nextPos = new IntVec2(0,0);
            for(IntRect r : gameArea.getMovingColliders().get(position.layer)) {
                nextPos.setX(position.nextX + Constant.TILE_SIZE / 2);
                nextPos.setY(position.nextY + Constant.TILE_SIZE / 2);
                if (!collider.collider.equals(r) && r.contains(nextPos)) return;
            }

            collider.collider.setX(position.nextX);
            collider.collider.setY(position.nextY);
            position.lastPixelStep = TimeUtils.millis();
            path.moving = true;
            path.startMoving = false;
        }

        // If moving, check whether next pixel step should take place
        if(!path.staticEntity && path.moving && TimeUtils.timeSinceMillis(position.lastPixelStep) >
                Constant.ONE_STEP_DURATION_PERSON && !path.talking) {
            switch(path.path.get(path.currentDir)) {
                case N: position.setY(position.getY() + 1);break;
                case W: position.setX(position.getX() - 1);break;
                case E: position.setX(position.getX() + 1);break;
                case S: position.setY(position.getY() - 1);break;
                default:path.stop();break;      // if stopping, count up stopping time
            }
            position.lastPixelStep = TimeUtils.millis();

            switch (path.path.get(path.currentDir)) {
                case N: if(position.getY() == position.nextY) path.moving = false;break;
                case E: if(position.getX() == position.nextX) path.moving = false;break;
                case W: if(position.getX() == position.nextX) path.moving = false;break;
                case S: if(position.getY() == position.nextY) path.moving = false;break;
                default: if(path.stopCounter==0) path.moving = false;break;
            }

            // Go on if finger is still on screen
            if(!path.moving) {
                path.next();
                path.startMoving = true;
            }
        }
    }
    /* ..................................................................... GETTERS & SETTERS .. */
}

File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/fwmengine/world/ecs/systems/PathSystem.kt
package de.limbusdev.guardianmonsters.fwmengine.world.ecs.systems

import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.EntitySystem
import com.badlogic.ashley.core.Family
import com.badlogic.ashley.utils.ImmutableArray
import com.badlogic.gdx.utils.TimeUtils

import de.limbusdev.guardianmonsters.Constant
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.ColliderComponent
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.InputComponent
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.PathComponent
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.components.PositionComponent
import de.limbusdev.guardianmonsters.utils.getComponent
import de.limbusdev.utils.geometry.IntVec2
import de.limbusdev.guardianmonsters.enums.SkyDirection


/**
 * Moves around entities with a [PathComponent] like persons, animals and so on
 * Created by georg on 30.11.15.
 */
class PathSystem(private val gameArea: GameArea) : EntitySystem()
{
    // --------------------------------------------------------------------------------------------- PROPERTIES
    companion object { const val TAG = "PathSystem" }

    private lateinit var entities: ImmutableArray<Entity>


    // --------------------------------------------------------------------------------------------- METHODS
    override fun addedToEngine(engine: Engine)
    {
        entities = engine.getEntitiesFor(Family.all(
                PositionComponent::class.java,
                ColliderComponent::class.java,
                PathComponent::class.java).exclude(
                InputComponent::class.java
        ).get())
    }

    override fun update(deltaTime: Float)
    {
        for (entity in entities)
        {
            val position = entity.getComponent<PositionComponent>()
            val collider = entity.getComponent<ColliderComponent>()
            val path     = entity.getComponent<PathComponent>()
            makeOneStep(position!!, path!!, collider!!)
            position.updateGridPosition()
        }
    }

    fun makeOneStep(position: PositionComponent, path: PathComponent, collider: ColliderComponent)
    {
        if (path.startMoving && !path.staticEntity)
        {
            // Define direction of movement
            when (path.path.get(path.currentDir))
            {
                SkyDirection.N ->
                {
                    position.nextX = position.x
                    position.nextY = position.y + Constant.TILE_SIZE
                }
                SkyDirection.W ->
                {
                    position.nextX = position.x - Constant.TILE_SIZE
                    position.nextY = position.y
                }
                SkyDirection.E ->
                {
                    position.nextX = position.x + Constant.TILE_SIZE
                    position.nextY = position.y
                }
                SkyDirection.S ->
                {
                    position.nextX = position.x
                    position.nextY = position.y - Constant.TILE_SIZE
                }
                else ->
                {
                    position.nextY = position.x
                    position.nextY = position.y
                }
            }

            // Check whether movement is possible or blocked by a collider
            val nextPos = IntVec2(0, 0)
            for (r in gameArea.dynamicColliders.get(position.layer))
            {
                nextPos.addOffset(Constant.TILE_SIZE/2)
                if (!collider.collider.equals(r) && r.contains(nextPos)) { return }
            }

            collider.collider += position
            position.lastPixelStep = TimeUtils.millis()
            path.moving = true
            path.startMoving = false
        }

        // If moving, check whether next pixel step should take place
        if (
                !path.staticEntity
                && path.moving
                && TimeUtils.timeSinceMillis(position.lastPixelStep) > Constant.ONE_STEP_DURATION_PERSON
                && !path.talking
        ) {
            position += when (path.path.get(path.currentDir))
            {
                SkyDirection.N -> IntVec2(0,1)
                SkyDirection.W -> IntVec2(-1,0)
                SkyDirection.E -> IntVec2(1,0)
                SkyDirection.S -> IntVec2(0,-1)
                else           -> { path.stop(); IntVec2() }
            }

            // if stopping, count up stopping time
            position.lastPixelStep = TimeUtils.millis()

            when (path.path.get(path.currentDir))
            {
                SkyDirection.N -> if (position.y == position.nextY) { path.moving = false }
                SkyDirection.E -> if (position.x == position.nextX) { path.moving = false }
                SkyDirection.W -> if (position.x == position.nextX) { path.moving = false }
                SkyDirection.S -> if (position.y == position.nextY) { path.moving = false }
                else           -> if (path.stopCounter == 0)        { path.moving = false }
            }

            // Go on if finger is still on screen
            if (!path.moving)
            {
                path.next()
                path.startMoving = true
            }
        }
    }
}
